File: php-api/src/Services/PortfolioService.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Services;

use PDO;

/**
 * Ported from Services/PortfolioService.cs. Backs the Org-Admin-only Portfolio Dashboard — the
 * first feature in this API where an Org Admin can pull data from projects they aren't necessarily
 * a *member* of (every other endpoint is gated by ProjectMemberMiddleware). Every method here takes
 * the caller's organisation id and independently re-validates every requested project id against it
 * before touching any data — a project id that doesn't belong to the caller's own org is silently
 * dropped from the result, never surfaced as a distinguishable error, so a client can't use this to
 * probe whether some other org's project id exists. `$validProjectIds` (re-derived from the DB,
 * never the raw request) is the only thing every query below is scoped by.
 */
final class PortfolioService
{
    public function __construct(private readonly PDO $db)
    {
    }

    public function listProjects(string $organisationId): array
    {
        $stmt = $this->db->prepare('SELECT "Id", "Name", "Key", "StartDate", "EndDate" FROM "Projects" WHERE "OrganisationId" = :orgId ORDER BY "Name"');
        $stmt->execute(['orgId' => $organisationId]);
        return array_map(static fn(array $p): array => [
            'id' => $p['Id'], 'name' => $p['Name'], 'key' => $p['Key'],
            'startDate' => $p['StartDate'], 'endDate' => $p['EndDate'],
        ], $stmt->fetchAll());
    }

    public function getAggregate(string $organisationId, array $requestedProjectIds): array
    {
        $validProjectIds = $this->validateProjectIds($organisationId, $requestedProjectIds);
        $orgUserCount = $this->countOrgUsers($organisationId);

        if (count($validProjectIds) === 0) {
            return [
                'members' => [], 'columns' => [], 'tasks' => [], 'releases' => [], 'risks' => [], 'decisions' => [],
                'startDate' => null, 'endDate' => null,
                'orgUserCount' => $orgUserCount, 'principleCount' => 0, 'objectiveCount' => 0,
                'documentCount' => 0, 'retrospectiveCount' => 0,
            ];
        }

        $members = [];
        $columns = [];
        $tasks = [];
        $releases = [];
        $risks = [];
        $decisions = [];
        $starts = [];
        $ends = [];
        $principleCount = 0;
        $objectiveCount = 0;
        $documentCount = 0;
        $retrospectiveCount = 0;

        $memberStmt = $this->db->prepare(<<<SQL
            SELECT m."Id", m."UserId", u."DisplayName", u."EmailAddress", m."Color", m."Role", m."ReportsToId"
            FROM "ProjectMembers" m JOIN "Users" u ON u."Id" = m."UserId"
            WHERE m."ProjectId" = :pid
        SQL);
        $columnStmt = $this->db->prepare('SELECT * FROM "Columns" WHERE "ProjectId" = :pid');
        $releaseStmt = $this->db->prepare('SELECT * FROM "Releases" WHERE "ProjectId" = :pid');
        $riskStmt = $this->db->prepare('SELECT r.*, p."Key" AS "ProjectKey" FROM "Risks" r JOIN "Projects" p ON p."Id" = r."ProjectId" WHERE r."ProjectId" = :pid');
        $decisionDocStmt = $this->db->prepare('SELECT "DocumentId" FROM "DecisionDocument" WHERE "DecisionId" = :id');
        $decisionRiskStmt = $this->db->prepare('SELECT "RiskId" FROM "DecisionRisk" WHERE "DecisionId" = :id');
        $decisionPrinStmt = $this->db->prepare('SELECT "PrincipleId" FROM "DecisionPrinciple" WHERE "DecisionId" = :id');
        $decisionObjStmt = $this->db->prepare('SELECT "ObjectiveId" FROM "DecisionObjective" WHERE "DecisionId" = :id');
        $decisionStmt = $this->db->prepare('SELECT * FROM "Decisions" WHERE "ProjectId" = :pid');
        $projectRangeStmt = $this->db->prepare('SELECT "StartDate", "EndDate" FROM "Projects" WHERE "Id" = :id');
        $principleCountStmt = $this->db->prepare('SELECT COUNT(*) FROM "Principles" WHERE "ProjectId" = :pid');
        $objectiveCountStmt = $this->db->prepare('SELECT COUNT(*) FROM "Objectives" WHERE "ProjectId" = :pid');
        $documentCountStmt = $this->db->prepare('SELECT COUNT(*) FROM "Documents" WHERE "ProjectId" = :pid');
        $retrospectiveCountStmt = $this->db->prepare('SELECT COUNT(*) FROM "Retrospectives" WHERE "ProjectId" = :pid');

        foreach ($validProjectIds as $projectId) {
            $memberStmt->execute(['pid' => $projectId]);
            foreach ($memberStmt->fetchAll() as $m) {
                $members[] = [
                    'id' => $m['Id'], 'userId' => $m['UserId'], 'displayName' => $m['DisplayName'],
                    'email' => $m['EmailAddress'], 'color' => $m['Color'], 'role' => $m['Role'], 'reportsToId' => $m['ReportsToId'],
                ];
            }

            $columnStmt->execute(['pid' => $projectId]);
            foreach ($columnStmt->fetchAll() as $c) {
                $columns[] = ['id' => $c['Id'], 'name' => $c['Name'], 'done' => (bool) $c['Done'], 'color' => $c['Color'], 'order' => (int) $c['Order']];
            }

            foreach (TaskService::fetchTaskDtos($this->db, $projectId) as $t) {
                $tasks[] = $t;
            }

            $releaseStmt->execute(['pid' => $projectId]);
            foreach ($releaseStmt->fetchAll() as $r) {
                $releases[] = [
                    'id' => $r['Id'], 'name' => $r['Name'], 'status' => $r['Status'], 'ownerId' => $r['OwnerId'],
                    'startDate' => $r['StartDate'], 'endDate' => $r['EndDate'],
                ];
            }

            $riskStmt->execute(['pid' => $projectId]);
            foreach ($riskStmt->fetchAll() as $r) {
                $risks[] = [
                    'id' => $r['Id'], 'key' => $r['Key'], 'title' => $r['Title'], 'description' => $r['Description'],
                    'likelihood' => (int) $r['Likelihood'], 'impact' => (int) $r['Impact'], 'mitigations' => $r['Mitigations'],
                    'ownerId' => $r['OwnerId'], 'taskId' => $r['TaskId'], 'status' => $r['Status'],
                    'dateToClose' => $r['DateToClose'], 'dateClosed' => $r['DateClosed'],
                    'projectId' => $r['ProjectId'], 'projectKey' => $r['ProjectKey'],
                ];
            }

            $decisionStmt->execute(['pid' => $projectId]);
            foreach ($decisionStmt->fetchAll() as $d) {
                $decisionDocStmt->execute(['id' => $d['Id']]);
                $decisionRiskStmt->execute(['id' => $d['Id']]);
                $decisionPrinStmt->execute(['id' => $d['Id']]);
                $decisionObjStmt->execute(['id' => $d['Id']]);
                $decisions[] = [
                    'id' => $d['Id'], 'key' => $d['Key'], 'title' => $d['Title'], 'description' => $d['Description'],
                    'type' => $d['Type'], 'status' => $d['Status'], 'outcome' => $d['Outcome'],
                    'ownerId' => $d['OwnerId'], 'approver' => $d['Approver'], 'taskId' => $d['TaskId'],
                    'documentIds' => $decisionDocStmt->fetchAll(PDO::FETCH_COLUMN),
                    'riskIds' => $decisionRiskStmt->fetchAll(PDO::FETCH_COLUMN),
                    'principleIds' => $decisionPrinStmt->fetchAll(PDO::FETCH_COLUMN),
                    'objectiveIds' => $decisionObjStmt->fetchAll(PDO::FETCH_COLUMN),
                ];
            }

            $projectRangeStmt->execute(['id' => $projectId]);
            $range = $projectRangeStmt->fetch();
            if ($range !== false) {
                if ($range['StartDate'] !== null) $starts[] = $range['StartDate'];
                if ($range['EndDate'] !== null) $ends[] = $range['EndDate'];
            }

            $principleCountStmt->execute(['pid' => $projectId]);
            $principleCount += (int) $principleCountStmt->fetchColumn();
            $objectiveCountStmt->execute(['pid' => $projectId]);
            $objectiveCount += (int) $objectiveCountStmt->fetchColumn();
            $documentCountStmt->execute(['pid' => $projectId]);
            $documentCount += (int) $documentCountStmt->fetchColumn();
            $retrospectiveCountStmt->execute(['pid' => $projectId]);
            $retrospectiveCount += (int) $retrospectiveCountStmt->fetchColumn();
        }

        sort($starts);
        rsort($ends);

        return [
            'members' => $members, 'columns' => $columns, 'tasks' => $tasks, 'releases' => $releases,
            'risks' => $risks, 'decisions' => $decisions,
            'startDate' => $starts[0] ?? null, 'endDate' => $ends[0] ?? null,
            'orgUserCount' => $orgUserCount, 'principleCount' => $principleCount, 'objectiveCount' => $objectiveCount,
            'documentCount' => $documentCount, 'retrospectiveCount' => $retrospectiveCount,
        ];
    }

    public function getActivity(string $organisationId, array $requestedProjectIds, string $start, string $end): array
    {
        $validProjectIds = $this->validateProjectIds($organisationId, $requestedProjectIds);
        if (count($validProjectIds) === 0) {
            return ['created' => [], 'edited' => [], 'done' => []];
        }

        $placeholders = implode(',', array_map(static fn(int $i): string => ":pid{$i}", array_keys($validProjectIds)));
        // Half-open [start, endExclusive) so the end date's own day is fully included — computed in
        // PHP since a bound string param can't have an interval added to it directly in SQL, same
        // convention as the .NET side.
        $endExclusive = (new \DateTimeImmutable($end))->modify('+1 day')->format('Y-m-d');
        $params = ['start' => $start, 'endExclusive' => $endExclusive];
        foreach ($validProjectIds as $i => $id) {
            $params["pid{$i}"] = $id;
        }

        $created = $this->fetchDailyBuckets(
            "SELECT date_trunc('day', \"DateCreated\") AS day, COUNT(*)::int AS n FROM \"Tasks\" " .
            "WHERE \"ProjectId\" IN ({$placeholders}) AND \"DateCreated\" >= :start AND \"DateCreated\" < :endExclusive " .
            "GROUP BY 1 ORDER BY 1",
            $params
        );
        $edited = $this->fetchDailyBuckets(
            "SELECT date_trunc('day', \"DateLastModified\") AS day, COUNT(*)::int AS n FROM \"Tasks\" " .
            "WHERE \"ProjectId\" IN ({$placeholders}) AND \"DateLastModified\" >= :start AND \"DateLastModified\" < :endExclusive " .
            "AND \"DateLastModified\" <> \"DateCreated\" GROUP BY 1 ORDER BY 1",
            $params
        );
        $done = $this->fetchDailyBuckets(
            "SELECT date_trunc('day', \"DateDone\") AS day, COUNT(*)::int AS n FROM \"Tasks\" " .
            "WHERE \"ProjectId\" IN ({$placeholders}) AND \"DateDone\" >= :start AND \"DateDone\" < :endExclusive " .
            "GROUP BY 1 ORDER BY 1",
            $params
        );

        return ['created' => $created, 'edited' => $edited, 'done' => $done];
    }

    private function fetchDailyBuckets(string $sql, array $params): array
    {
        $stmt = $this->db->prepare($sql);
        $stmt->execute($params);
        return array_map(static fn(array $row): array => [
            'date' => substr((string) $row['day'], 0, 10),
            'count' => (int) $row['n'],
        ], $stmt->fetchAll());
    }

    /**
     * Backs the Timeline chart's click-to-edit modal and drag-to-schedule bars. Its own endpoint
     * rather than reusing ProjectsController's PUT /projects/{id} — that one requires
     * ProjectMemberMiddleware, which an Org Admin scheduling a project they don't personally belong
     * to would fail. OrgAdmin + org-ownership check only, same as every other method here. Either
     * date may be null to clear it (reverting the project back to the "no dates" state).
     */
    public function updateProjectDates(string $organisationId, string $projectId, ?string $startDate, ?string $endDate): bool
    {
        $stmt = $this->db->prepare('UPDATE "Projects" SET "StartDate" = :start, "EndDate" = :end, "DateLastModified" = now() WHERE "Id" = :id AND "OrganisationId" = :orgId');
        $stmt->execute(['start' => $startDate, 'end' => $endDate, 'id' => $projectId, 'orgId' => $organisationId]);
        return $stmt->rowCount() > 0;
    }

    /** The one place a client-supplied project id list is trusted at all: re-derived against the
     * caller's own organisation, so every subsequent query only ever touches project ids proven to
     * belong to the caller's org. */
    private function validateProjectIds(string $organisationId, array $requestedProjectIds): array
    {
        $requestedProjectIds = array_values(array_filter($requestedProjectIds, static fn($id): bool => is_string($id) && $id !== ''));
        if (count($requestedProjectIds) === 0) {
            return [];
        }

        $placeholders = implode(',', array_map(static fn(int $i): string => ":id{$i}", array_keys($requestedProjectIds)));
        $stmt = $this->db->prepare("SELECT \"Id\" FROM \"Projects\" WHERE \"OrganisationId\" = :orgId AND \"Id\" IN ({$placeholders})");
        $params = ['orgId' => $organisationId];
        foreach ($requestedProjectIds as $i => $id) {
            $params["id{$i}"] = $id;
        }
        $stmt->execute($params);
        return $stmt->fetchAll(PDO::FETCH_COLUMN);
    }

    private function countOrgUsers(string $organisationId): int
    {
        $stmt = $this->db->prepare('SELECT COUNT(*) FROM "Users" WHERE "OrganisationId" = :orgId AND "IsActive" = true');
        $stmt->execute(['orgId' => $organisationId]);
        return (int) $stmt->fetchColumn();
    }
}

File: api/Enkl.Api/Services/PortfolioService.cs
using Enkl.Api.Data;
using Enkl.Api.Dtos;
using Microsoft.EntityFrameworkCore;

namespace Enkl.Api.Services;

/// <summary>
/// Backs the Org-Admin-only Portfolio Dashboard — the first feature in this API where an Org Admin
/// can pull data from projects they aren't necessarily a *member* of (every other endpoint is
/// [Authorize(Policy = "ProjectMember")]). Every method here takes the caller's OrganisationId and
/// independently re-validates every requested project id against it before touching any data — a
/// project id that doesn't belong to the caller's own org is silently dropped from the result, never
/// surfaced as a distinguishable error, so a client can't use this to probe whether some other org's
/// project id exists. `validProjectIds` (re-derived from the DB, never the raw request) is the only
/// thing every query below is scoped by.
/// </summary>
public class PortfolioService
{
    private readonly AppDbContext _db;

    public PortfolioService(AppDbContext db)
    {
        _db = db;
    }

    public async Task<List<PortfolioProjectDto>> ListProjectsAsync(Guid organisationId)
    {
        return await _db.Projects
            .Where(p => p.OrganisationId == organisationId)
            .OrderBy(p => p.Name)
            .Select(p => new PortfolioProjectDto(p.Id, p.Name, p.Key, p.StartDate, p.EndDate))
            .ToListAsync();
    }

    public async Task<PortfolioAggregateDto> GetAggregateAsync(Guid organisationId, List<Guid> requestedProjectIds)
    {
        var validProjectIds = await ValidateProjectIdsAsync(organisationId, requestedProjectIds);
        var orgUserCount = await CountOrgUsersAsync(organisationId);

        if (validProjectIds.Count == 0)
        {
            return new PortfolioAggregateDto(
                new(), new(), new(), new(), new(), new(),
                StartDate: null, EndDate: null,
                OrgUserCount: orgUserCount, PrincipleCount: 0, ObjectiveCount: 0, DocumentCount: 0, RetrospectiveCount: 0);
        }

        var members = await _db.ProjectMembers
            .Where(m => validProjectIds.Contains(m.ProjectId))
            .Select(m => new MemberDto(m.Id, m.UserId, m.User.DisplayName, m.User.EmailAddress, m.Color, m.Role, m.ReportsToId))
            .ToListAsync();

        var columns = await _db.Columns
            .Where(c => validProjectIds.Contains(c.ProjectId))
            .Select(c => new ColumnDto(c.Id, c.Name, c.Done, c.Color, c.Order))
            .ToListAsync();

        // Materialize entities first, then map in-memory (ProjectService.ToTaskDto isn't EF
        // Core-translatable) — same two-step shape ProjectService.GetProjectDetailAsync itself uses.
        var taskEntities = await _db.Tasks
            .Where(t => validProjectIds.Contains(t.ProjectId))
            .Include(t => t.Dependencies)
            .Include(t => t.AuditLog)
            .ToListAsync();
        var tasks = taskEntities.Select(ProjectService.ToTaskDto).ToList();

        var releases = await _db.Releases
            .Where(r => validProjectIds.Contains(r.ProjectId))
            .Select(r => new ReleaseDto(r.Id, r.Name, r.Status, r.OwnerId, r.StartDate, r.EndDate))
            .ToListAsync();

        var risks = await _db.Risks
            .Where(r => validProjectIds.Contains(r.ProjectId))
            .Select(r => new PortfolioRiskDto(
                r.Id, r.Key, r.Title, r.Description, r.Likelihood, r.Impact, r.Mitigations,
                r.OwnerId, r.TaskId, r.Status, r.DateToClose, r.DateClosed,
                r.ProjectId, r.Project.Key))
            .ToListAsync();

        var decisions = await _db.Decisions
            .Where(d => validProjectIds.Contains(d.ProjectId))
            .Select(d => new DecisionDto(
                d.Id, d.Key, d.Title, d.Description, d.Type, d.Status, d.Outcome, d.OwnerId, d.Approver, d.TaskId,
                d.Documents.Select(x => x.DocumentId).ToList(), d.Risks.Select(x => x.RiskId).ToList(),
                d.Principles.Select(x => x.PrincipleId).ToList(), d.Objectives.Select(x => x.ObjectiveId).ToList()))
            .ToListAsync();

        var projectRanges = await _db.Projects
            .Where(p => validProjectIds.Contains(p.Id))
            .Select(p => new { p.StartDate, p.EndDate })
            .ToListAsync();
        var starts = projectRanges.Where(p => p.StartDate.HasValue).Select(p => p.StartDate!.Value).ToList();
        var ends = projectRanges.Where(p => p.EndDate.HasValue).Select(p => p.EndDate!.Value).ToList();

        var principleCount = await _db.Principles.CountAsync(p => validProjectIds.Contains(p.ProjectId));
        var objectiveCount = await _db.Objectives.CountAsync(o => validProjectIds.Contains(o.ProjectId));
        var documentCount = await _db.Documents.CountAsync(d => validProjectIds.Contains(d.ProjectId));
        var retrospectiveCount = await _db.Retrospectives.CountAsync(r => validProjectIds.Contains(r.ProjectId));

        return new PortfolioAggregateDto(
            members, columns, tasks, releases, risks, decisions,
            StartDate: starts.Count > 0 ? starts.Min() : null,
            EndDate: ends.Count > 0 ? ends.Max() : null,
            OrgUserCount: orgUserCount,
            PrincipleCount: principleCount, ObjectiveCount: objectiveCount,
            DocumentCount: documentCount, RetrospectiveCount: retrospectiveCount);
    }

    public async Task<PortfolioActivityDto> GetActivityAsync(Guid organisationId, List<Guid> requestedProjectIds, DateOnly start, DateOnly end)
    {
        var validProjectIds = await ValidateProjectIdsAsync(organisationId, requestedProjectIds);
        if (validProjectIds.Count == 0) return new PortfolioActivityDto(new(), new(), new());

        // DateTime.Kind must be Utc — Tasks' DateCreated/DateLastModified/DateDone columns are
        // `timestamp with time zone`, and Npgsql refuses to bind a Kind=Unspecified DateTime against them.
        var startDt = DateTime.SpecifyKind(start.ToDateTime(TimeOnly.MinValue), DateTimeKind.Utc);
        // Half-open [start, endExclusive) so the end date's own day is fully included, same
        // convention vendor-portal's own /dashboard/activity endpoint uses.
        var endExclusiveDt = DateTime.SpecifyKind(end.AddDays(1).ToDateTime(TimeOnly.MinValue), DateTimeKind.Utc);

        var created = await _db.Tasks
            .Where(t => validProjectIds.Contains(t.ProjectId) && t.DateCreated >= startDt && t.DateCreated < endExclusiveDt)
            .GroupBy(t => t.DateCreated.Date)
            .Select(g => new PortfolioActivityPointDto(DateOnly.FromDateTime(g.Key), g.Count()))
            .ToListAsync();

        // "Edited" excludes a task's own creation timestamp (DateLastModified == DateCreated at
        // insert time) — otherwise every created task would also silently count as "edited" on the
        // same day, matching vendor-portal's identical exclusion.
        var edited = await _db.Tasks
            .Where(t => validProjectIds.Contains(t.ProjectId) && t.DateLastModified >= startDt && t.DateLastModified < endExclusiveDt && t.DateLastModified != t.DateCreated)
            .GroupBy(t => t.DateLastModified.Date)
            .Select(g => new PortfolioActivityPointDto(DateOnly.FromDateTime(g.Key), g.Count()))
            .ToListAsync();

        var done = await _db.Tasks
            .Where(t => validProjectIds.Contains(t.ProjectId) && t.DateDone != null && t.DateDone >= startDt && t.DateDone < endExclusiveDt)
            .GroupBy(t => t.DateDone!.Value.Date)
            .Select(g => new PortfolioActivityPointDto(DateOnly.FromDateTime(g.Key), g.Count()))
            .ToListAsync();

        return new PortfolioActivityDto(
            created.OrderBy(p => p.Date).ToList(),
            edited.OrderBy(p => p.Date).ToList(),
            done.OrderBy(p => p.Date).ToList());
    }

    /// <summary>
    /// Backs the Timeline chart's click-to-edit modal and drag-to-schedule bars. Deliberately its own
    /// endpoint rather than reusing ProjectsController's PUT /api/projects/{id} — that one requires
    /// ProjectMember (see ProjectsController.cs), which an Org Admin scheduling a project they don't
    /// personally belong to would fail. OrgAdmin + org-ownership check only, same as every other
    /// method here. Returns false (not found / wrong org) without revealing which — same
    /// no-enumeration-oracle stance as ValidateProjectIdsAsync below.
    /// </summary>
    public async Task<bool> UpdateProjectDatesAsync(Guid organisationId, Guid projectId, DateOnly? startDate, DateOnly? endDate)
    {
        var project = await _db.Projects.FirstOrDefaultAsync(p => p.Id == projectId && p.OrganisationId == organisationId);
        if (project is null) return false;

        project.StartDate = startDate;
        project.EndDate = endDate;
        project.DateLastModified = DateTime.UtcNow;
        await _db.SaveChangesAsync();
        return true;
    }

    /// <summary>The one place a client-supplied project id list is trusted at all: re-derived
    /// against the caller's own OrganisationId, so every subsequent query in this class only ever
    /// touches project ids proven to belong to the caller's org.</summary>
    private async Task<List<Guid>> ValidateProjectIdsAsync(Guid organisationId, List<Guid> requestedProjectIds)
    {
        if (requestedProjectIds.Count == 0) return new List<Guid>();
        return await _db.Projects
            .Where(p => requestedProjectIds.Contains(p.Id) && p.OrganisationId == organisationId)
            .Select(p => p.Id)
            .ToListAsync();
    }

    private async Task<int> CountOrgUsersAsync(Guid organisationId) =>
        await _db.Users.CountAsync(u => u.OrganisationId == organisationId && u.IsActive);
}
